src/sync/mpsc/tip.rs

```diff
@@ -0,0 +1,135 @@
+use std::sync::{Arc, Mutex};
+
+use task::{self, Task};
+use {Sink, Stream, AsyncSink, Async, Poll, StartSend};
+
+/// Tip is very similar to unbounded channel but only stores last value sent
+///
+/// I.e. if you want to send some value between from producer to a consumer
+/// and if consumer is slow it should may skip old values, the tip is
+/// a structure for the task.
+
+/// The transmission end of a channel which is used to send values
+///
+/// If the receiver is not fast enough only the last value is preserved and
+/// other ones are discarded.
+#[derive(Debug)]
+pub struct TipSender<T> {
+    inner: Arc<Mutex<Inner<T>>>,
+}
+
+/// The receiving end of a channel which preserves only the last value
+#[derive(Debug)]
+pub struct TipReceiver<T> {
+    inner: Arc<Mutex<Inner<T>>>,
+}
+
+#[derive(Debug)]
+struct Inner<T> {
+    closed: bool,
+    value: Option<T>,
+    task: Option<Task>,
+}
+
+trait AssertKindsSender: Send + Sync + Clone {}
+impl AssertKindsSender for TipSender<u32> {}
+
+trait AssertKindsReceiver: Send + Sync {}
+impl AssertKindsReceiver for TipReceiver<u32> {}
+
+impl<T> TipSender<T> {
+    /// Sets the new new value of the stream and notifies the consumer if any
+    pub fn set(&self, value: T) {
+        // Do this step first so that the lock is dropped when
+        // `unpark` is called
+        let task = {
+            let mut inner = self.inner.lock().unwrap();
+            inner.value = Some(value);
+            inner.task.take()
+        };
+        if let Some(task) = task {
+            task.notify();
+        }
+    }
+}
+
+impl<T> Sink for TipSender<T> {
+    type SinkItem = T;
+    type SinkError = ();  // actually void
+    fn start_send(&mut self, item: T) -> StartSend<T, ()> {
+        self.set(item);
+        Ok(AsyncSink::Ready)
+    }
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        Ok(Async::Ready(()))
+    }
+    fn close(&mut self) -> Poll<(), Self::SinkError> {
+        let task = {
+            let mut inner = self.inner.lock().unwrap();
+            inner.closed = true;
+            inner.task.take()
+        };
+        if let Some(task) = task {
+            task.notify();
+        }
+        Ok(Async::Ready(()))
+    }
+}
+
+impl<T> Drop for TipSender<T> {
+    fn drop(&mut self) {
+        // if there are only two strong counts: this one and receiver we
+        // should notify receiver that stream is closed. If these two
+        // references are just two senders the wake up isn't problematic.
+        if Arc::strong_count(&self.inner) == 2 {
```

Comment:
Does this race if two senders in two threads drop at the same time? Wouldn't both check this, see 3, and do nothing?